File: SMO/SmoBrowser/CS/SmoBrowser/SmoBrowser.cs
/*============================================================================
  File:    SmoBrowser.cs 

  Summary: Implements a sample SMO browser utility in C#.

  Date:    August 16, 2005
------------------------------------------------------------------------------
  This file is part of the Microsoft SQL Server Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
============================================================================*/

#region Using directives

using System;
using System.Collections;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Globalization;

using Microsoft.SqlServer.Management.Common;
using Microsoft.SqlServer.Management.Smo;
using Microsoft.SqlServer.Management.Smo.Wmi;
using Microsoft.SqlServer.MessageBox;

#endregion

namespace Microsoft.Samples.SqlServer
{
    partial class SmoBrowser : Form
    {
        // Use the Server object to connect to a specific server
        private Server SqlServerSelection;

        public SmoBrowser()
        {
            InitializeComponent();
        }

        private void SmoBrowser_Load(object sender, EventArgs e)
        {
            ServerConnection ServerConn = new ServerConnection();
            ServerConnect scForm;
            DialogResult dr;

            // Display the main window first
            this.Show();
            Application.DoEvents();
            
            ServerConn = new ServerConnection();
            scForm = new ServerConnect(ServerConn);
            dr = scForm.ShowDialog(this);
            if ((dr == DialogResult.OK) &&
                (ServerConn.SqlConnectionObject.State == ConnectionState.Open))
            {
                SqlServerSelection = new Server(ServerConn);
                if (SqlServerSelection != null)
                {
                    this.Text = Properties.Resources.AppTitle
                        + SqlServerSelection.Name;
                }
            }
            else
            {
                this.Close();
            }

            if (SqlServerSelection != null)
            {
                objectTreeView.BeginUpdate();

                // By default create Server and Managed Computer nodes
                objectTreeView.Nodes.Add(CreateNode(
                    String.Format(CultureInfo.CurrentUICulture,
                    Properties.Resources.SQLServerNodeName,
                    SqlServerSelection.Name),
                    SqlServerSelection));
                objectTreeView.Nodes.Add(CreateNode(
                    String.Format(CultureInfo.CurrentUICulture,
                    Properties.Resources.ManagedComputerNodeName,
                    SqlServerSelection.Name),
                    new ManagedComputer(SqlServerSelection.Name)));

                objectTreeView.EndUpdate();
            }
        }

        private void exitToolStripMenuItem_Click(object sender, EventArgs e)
        {
            Close();
            Application.Exit();
        }

        private void OnBeforeExpand(object sender, TreeViewCancelEventArgs e)
        {
            try
            {
                if (e.Node.Nodes.Count == 1
                    && e.Node.Nodes[0].Text == Properties.Resources.NodePlaceHolder)
                {
                    e.Node.Nodes.RemoveAt(0);
                    if (e.Node.Tag is ICollection)
                    {
                        PopulateCollectionItems(e.Node);
                    }
                    else
                    {
                        PopulateExpandableProperties(e.Node);
                    }
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
        }

        private void AfterSelect(object sender, TreeViewEventArgs e)
        {
            Cursor oldCursor = this.Cursor;
            this.Cursor = Cursors.WaitCursor;

            try
            {
                this.propertyGrid1.SelectedObject = e.Node.Tag;
                this.textBox1.Text = GetScriptText(e.Node.Tag as IScriptable);
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                this.Cursor = oldCursor;
            }
        }

        private static string GetScriptText(IScriptable scriptable)
        {
            if (scriptable == null)
            {
                return String.Empty;
            }

            StringBuilder stringBuilder = new StringBuilder();
            StringCollection batches;
            try
            {
                batches = scriptable.Script();
            }
            catch {
                return string.Empty;
            }
            foreach (string batch in batches)
            {
                stringBuilder.AppendFormat(@"{0}\r\nGO\r\n", batch);
            }

            return stringBuilder.ToString();
        }

        private static TreeNode CreateNode(object item)
        {
            string name = null;
            PropertyDescriptor property;

            foreach (string propertyName in new string[] { @"Name", @"Urn" })
            {
                property = TypeDescriptor.GetProperties(item)[propertyName];
                if (property != null)
                {
                    name = property.GetValue(item).ToString();
                    break;
                }
            }

            if (name == null)
            {
                name = item.ToString();
            }

            return CreateNode(name, item);
        }

        private static TreeNode CreateNode(string name, object item)
        {
            TreeNode node = new TreeNode(name);

            node.Tag = item;
            if (item is ICollection || HasExpandableProperties(item))
            {
                node.Nodes.Add(Properties.Resources.NodePlaceHolder);
            }

            return node;
        }

        private static bool HasExpandableProperties(object item)
        {
            foreach (PropertyDescriptor property in TypeDescriptor.GetProperties(item))
            {
                if (IsExpandableProperty(property))
                {
                    return true;
                }
            }

            return false;
        }

        private static bool IsCollection(PropertyDescriptor property)
        {
            foreach (Type type in property.PropertyType.GetInterfaces())
            {
                if (type == typeof(ICollection))
                {
                    if (property.PropertyType.IsArray)
                    {
                        break;
                    }

                    if (property.Name == @"Properties")
                    {
                        break;
                    }

                    return true;
                }
            }

            return false;
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
        private void PopulateExpandableProperties(TreeNode node)
        {
            object value;
            Cursor oldCursor = this.Cursor;
            this.Cursor = Cursors.WaitCursor;

            try
            {
                foreach (PropertyDescriptor property in TypeDescriptor.GetProperties(node.Tag))
                {
                    if (!IsExpandableProperty(property))
                    {
                        continue;
                    }

                    try
                    {
                        value = property.GetValue(node.Tag);
                    }
                    catch (Exception ex)
                    {
                        value = String.Format(CultureInfo.CurrentUICulture,
                            Properties.Resources.ExceptionString,
                            ex.GetType(), ex.Message);
                    }

                    node.Nodes.Add(CreateNode(property.Name, value));
                }
            }
            finally
            {
                this.Cursor = oldCursor;
            }
        }

        private static bool IsExpandableProperty(PropertyDescriptor property)
        {
            if (!IsExpandablePropertyType(property.PropertyType))
            {
                return false;
            }

            if (property.PropertyType.IsArray)
            {
                if (!IsExpandablePropertyType(property.PropertyType.GetElementType()))
                {
                    return false;
                }
            }

            if (property.Name == @"Urn" ||
                property.Name == @"UserData" ||
                property.Name == @"Properties" ||
                property.Name == @"ExtendedProperties" ||
                property.Name == @"Parent" ||
                property.Name == @"Events")
            {
                return false;
            }

            return true;
        }

        private static bool IsExpandablePropertyType(Type type)
        {
            if (Type.GetTypeCode(type) != TypeCode.Object)
            {
                return false;
            }

            if (type == typeof(Guid) || type == typeof(DateTime))
            {
                return false;
            }

            return true;
        }

        private void PopulateCollectionItems(TreeNode node)
        {
            Cursor oldCursor = this.Cursor;
            this.Cursor = Cursors.WaitCursor;

            try
            {
                foreach (object item in (ICollection)node.Tag)
                {
                    node.Nodes.Add(CreateNode(item));
                }
            }
            finally
            {
                this.Cursor = oldCursor;
            }
        }
    }
}

File: SMO/SmoBrowser/VB/SmoBrowser/SmoBrowser.vb
'============================================================================
'  File:    SmoBrowser.vb 
'
'  Summary: Implements a sample SMO browser utility in VB.NET.
'
'  Date:    August 16, 2005
'------------------------------------------------------------------------------
'  This file is part of the Microsoft SQL Server Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
'
'  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'============================================================================

#Region "Using directives"
Imports System
Imports System.Collections
Imports System.Collections.Specialized
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Globalization

Imports Microsoft.SqlServer.Management.Common
Imports Microsoft.SqlServer.Management.Smo
Imports Microsoft.SqlServer.Management.Smo.Wmi
Imports Microsoft.SqlServer.MessageBox
#End Region

Partial Class SmoBrowser 'The Partial modifier is only required on one class definition per project.
    Inherits Form
    ' Use the Server object to connect to a specific server
    Private SqlServerSelection As Server

    Public Sub New()
        InitializeComponent()
    End Sub

    Private Sub SmoBrowser_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim ServerConn As New ServerConnection()
        Dim scForm As ServerConnect
        Dim dr As DialogResult

        ' Display the main window first
        Me.Show()
        Application.DoEvents()

        ServerConn = New ServerConnection()
        scForm = New ServerConnect(ServerConn)
        dr = scForm.ShowDialog(Me)
        If dr = Windows.Forms.DialogResult.OK AndAlso ServerConn.SqlConnectionObject.State = ConnectionState.Open Then
            SqlServerSelection = New Server(ServerConn)
            If Not (SqlServerSelection Is Nothing) Then
                Me.Text = My.Resources.AppTitle + SqlServerSelection.Name
            End If
        Else
            Me.Close()
        End If

        If Not (SqlServerSelection Is Nothing) Then
            ' By default create Server and Managed Computer nodes
            objectTreeView.Nodes.Add(CreateNode(String.Format(CultureInfo.CurrentUICulture, My.Resources.SQLServerNodeName, SqlServerSelection.Name), SqlServerSelection))
            objectTreeView.Nodes.Add(CreateNode(String.Format(CultureInfo.CurrentUICulture, My.Resources.ManagedComputerNodeName, SqlServerSelection.Name), New ManagedComputer(SqlServerSelection.Name)))
        End If
    End Sub

    Private Sub exitToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles exitToolStripMenuItem.Click
        Close()
        Application.Exit()
    End Sub

    Private Sub OnBeforeExpand(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TreeViewCancelEventArgs) Handles objectTreeView.BeforeExpand
        Try
            If e.Node.Nodes.Count = 1 AndAlso e.Node.Nodes(0).Text = My.Resources.NodePlaceHolder Then
                e.Node.Nodes.RemoveAt(0)
                If TypeOf e.Node.Tag Is ICollection Then
                    PopulateCollectionItems(e.Node)
                Else
                    PopulateExpandableProperties(e.Node)
                End If
            End If
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        End Try
    End Sub

    Private Sub AfterSelect(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TreeViewEventArgs) Handles objectTreeView.AfterSelect
        Dim oldCursor As Cursor = Me.Cursor
        Me.Cursor = Cursors.WaitCursor

        Try
            Me.propertyGrid1.SelectedObject = e.Node.Tag
            If TypeOf e.Node.Tag Is IScriptable Then
                Me.textBox1.Text = GetScriptText(CType(e.Node.Tag, IScriptable))
            Else
                Me.textBox1.Text = String.Empty
            End If

        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            Me.Cursor = oldCursor
        End Try
    End Sub

    Private Shared Function GetScriptText(ByVal scriptable As IScriptable) As String
        If scriptable Is Nothing Then
            Return String.Empty
        End If

        Dim stringBuilder As New StringBuilder()
        Dim batches As StringCollection
        Try
            batches = scriptable.Script()
        Catch ex As Exception
            Return String.Empty
        End Try
        For Each batch As String In batches
            stringBuilder.AppendFormat("{0}" & vbCrLf & "GO" & vbCrLf, batch)
        Next

        Return stringBuilder.ToString()
    End Function

    Private Overloads Shared Function CreateNode(ByVal item As Object) As TreeNode
        Dim name As String = Nothing
        Dim [property] As PropertyDescriptor

        For Each propertyName As String In New String() {"Name", "Urn"}
            [property] = TypeDescriptor.GetProperties(item)(propertyName)
            If Not ([property] Is Nothing) Then
                name = [property].GetValue(item).ToString()
                Exit For
            End If
        Next

        If name Is Nothing Then
            name = item.ToString()
        End If

        Return CreateNode(name, item)
    End Function

    Private Overloads Shared Function CreateNode(ByVal name As String, ByVal item As Object) As TreeNode
        Dim node As New TreeNode(name)

        node.Tag = item
        If TypeOf item Is ICollection OrElse HasExpandableProperties(item) Then
            node.Nodes.Add(My.Resources.NodePlaceHolder)
        End If

        Return node
    End Function

    Private Shared Function HasExpandableProperties(ByVal item As Object) As Boolean
        For Each [property] As PropertyDescriptor In TypeDescriptor.GetProperties(item)
            If IsExpandableProperty([property]) = True Then
                Return True
            End If
        Next

        Return False
    End Function

    Private Shared Function IsCollection(ByVal [property] As PropertyDescriptor) As Boolean
        For Each typ As Type In [property].PropertyType.GetInterfaces()
            If TypeOf typ Is ICollection Then
                If [property].PropertyType.IsArray Then
                    Exit For
                End If

                If [property].Name = "Properties" Then
                    Exit For
                End If

                Return True
            End If
        Next

        Return False
    End Function

    <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")> Private Sub PopulateExpandableProperties(ByVal node As TreeNode)
        Dim value As Object
        Dim oldCursor As Cursor = Me.Cursor
        Me.Cursor = Cursors.WaitCursor

        Try
            For Each [property] As PropertyDescriptor In TypeDescriptor.GetProperties(node.Tag)
                If IsExpandableProperty([property]) = False Then
                    Continue For
                End If

                Try
                    value = [property].GetValue(node.Tag)
                Catch ex As Exception
                    value = String.Format(CultureInfo.CurrentUICulture, _
                        My.Resources.ExceptionString, ex.GetType(), ex.Message)
                End Try

                node.Nodes.Add(CreateNode([property].Name, value))
            Next
        Finally
            Me.Cursor = oldCursor
        End Try
    End Sub

    Private Shared Function IsExpandableProperty(ByVal [property] As PropertyDescriptor) As Boolean
        If Not IsExpandablePropertyType([property].PropertyType) Then
            Return False
        End If

        If [property].PropertyType.IsArray Then
            If Not IsExpandablePropertyType([property].PropertyType.GetElementType()) Then
                Return False
            End If
        End If

        If [property].Name = "Urn" _
            OrElse [property].Name = "UserData" _
            OrElse [property].Name = "Properties" _
            OrElse [property].Name = "ExtendedProperties" _
            OrElse [property].Name = "Parent" _
            OrElse [property].Name = "Events" Then
            Return False
        End If

        Return True
    End Function

    Private Shared Function IsExpandablePropertyType(ByVal typ As Type) As Boolean
        If Type.GetTypeCode(typ) <> TypeCode.Object Then
            Return False
        End If

        If typ Is GetType(Guid) OrElse typ Is GetType(DateTime) Then
            Return False
        End If

        Return True
    End Function

    Private Sub PopulateCollectionItems(ByVal node As TreeNode)
        Dim oldCursor As Cursor = Me.Cursor
        Me.Cursor = Cursors.WaitCursor

        Try
            For Each item As Object In CType(node.Tag, ICollection)
                node.Nodes.Add(CreateNode(item))
            Next
        Finally
            Me.Cursor = oldCursor
        End Try
    End Sub
End Class
